views.go

```diff
@@ -698,75 +538,54 @@ func (lk *logKeeper) viewBucketLogs(r *http.Request, buildID string, testID stri
-	wg.Add(2)
+	wg.Add(3)
+	go func() {
+		defer recovery.LogStackTraceAndContinue("finding build from bucket")
+		defer wg.Done()
+
+		build, buildErr = lk.opts.Bucket.FindBuildByID(r.Context(), buildID)
+	}()
 	go func() {
 		defer recovery.LogStackTraceAndContinue("finding test for build from bucket")
 		defer wg.Done()
 
 		if testID == "" {
 			return
 		}
-		test, testErr = lk.opts.Bucket.FindTestByID(backgroundContext, buildID, testID)
+		test, testErr = lk.opts.Bucket.FindTestByID(r.Context(), buildID, testID)
 	}()
 	go func() {
 		defer recovery.LogStackTraceAndContinue("downloading log lines from bucket")
 		defer wg.Done()
 
-		logLines, logLinesErr = lk.opts.Bucket.DownloadLogLines(backgroundContext, buildID, testID)
+		logLines, logLinesErr = lk.opts.Bucket.DownloadLogLines(r.Context(), buildID, testID)
 	}()
+	wg.Wait()
 
-	build, buildErr = lk.opts.Bucket.FindBuildByID(r.Context(), buildID)
-	if pail.IsKeyNotFoundError(buildErr) {
-		backgroundCancel()
-		if shouldFallBack {
-			grip.Info(message.Fields{
-				"message":   "fallback to db",
-				"operation": "viewBucketLogs",
-				"build_id":  buildID,
-				"test_id":   testID,
-			})
-			if testID == "" {
-				return lk.viewAllDBLogs(r, buildID)
-			} else {
-				return lk.viewDBTestLogs(r, buildID, testID)
-			}
-		} else {
-			return nil, &apiError{Err: "finding build", code: http.StatusNotFound}
-		}
-	}
 	if buildErr != nil {
-		backgroundCancel()
 		lk.logErrorf(r, "finding build '%s': %v", buildID, buildErr)
-		return nil, &apiError{Err: "finding build", code: http.StatusInternalServerError}
+		return nil, &apiError{Err: "finding build", code: http.StatusInternalServerError}, false
 	}
-
-	wg.Wait()
-	if testID != "" && pail.IsKeyNotFoundError(testErr) {
-		backgroundCancel()
-		return nil, &apiError{Err: "test not found", code: http.StatusNotFound}
+	if build == nil {
+		return nil, &apiError{Err: "build not found", code: http.StatusNotFound}, true
 	}
 	if testErr != nil {
 		lk.logErrorf(r, "finding test '%s' for build '%s': %v", testID, buildID, testErr)
-		backgroundCancel()
-		return nil, &apiError{Err: "finding test", code: http.StatusInternalServerError}
+		return nil, &apiError{Err: "finding test", code: http.StatusInternalServerError}, false
+	}
+	if testID != "" && test == nil {
+		return nil, &apiError{Err: "test not found", code: http.StatusNotFound}, false
 	}
 	if logLinesErr != nil {
 		lk.logErrorf(r, "downloading logs for build '%s': %v", buildID, logLinesErr)
-		backgroundCancel()
-		return nil, &apiError{Err: "downloading logs", code: http.StatusInternalServerError}
+		return nil, &apiError{Err: "downloading logs", code: http.StatusInternalServerError}, false
 	}
 
 	return &logFetchResponse{
-		logLines:      logLines,
-		build:         build,
-		test:          test,
-		contextCancel: backgroundCancel,
-	}, nil
+		logLines: logLines,
+		build:    build,
+		test:     test,
+	}, nil, false
```

Comment:
should the contextCancel be removed from the `logFetchResponse` struct too?

Reply:
Yes, good catch!

Reply:
Done.